### ISSUE TITLE
Go To Definition/Declaration feature request

### ISSUE THREAD

**Issue by terminalObserver** [geany/geany#3379]
With a valid tag file loaded;
I would like to improve the "Go To..." feature by having geany open the file of the reference in Ctags.
This saves user from manually locating the #included file, opening it, and then viewing the code.

**Comment by elextr** [comment]
It does ... if:

1. the tag is found in the loaded tags
2. the tags file contains a filename, Geany generated global tags files do not
3. the filename points to a file thats found by `realpath(filename)`, remember Geany does not know about -I include commands.

I think if the tags file is generated by uctags it will have the filename, and maybe Geany will read that (@techee does it?).

**Comment by techee** [comment]
Currently, Geany ignores filename in global tags - to be precise, Geany's internal tag file format doesn't support file name at all and the filename field from the ctags file format is ignored right now.

If we wanted to support this feature, we'd have to modify quite a bit of the TM code because right now the presence of filename is used to distinguish global tags from non-global ones:
```
if (tag->file) {
    // this is not a global tag
} else {
    // this is a global tag
}
```

To workaround the original problem, @terminalObserver could use some plugin such as ProjectOrganizer where it's possible to add external directories to a project (containing e.g. the include files he wants to open) and then all the features like goto tag definition/declaration work.

**Comment by terminalObserver** [comment]
Thanks,
I am already using the ProjectOrganzier plug-in, with the included files present in the project directory and visible in the organizer. The "Go To..." functions do not work unless the document is opened in the notebook.

**Comment by techee** [comment]
Then I assume that:

1. Either you didn't set "Index all project files" to "Yes" under Project->Properties->ProjectOrganizer tab,
2. or the include files you want to open are outside the project directory in which case you have to add the directory containing them using the toolbar icon with the "folder and plus" in the ProjectOrganizer sidebar,
3. or both of the above.

**Comment by terminalObserver** [comment]
Thanks!

Project->Properties->ProjectOrganizer was set to "Auto" ; It now works as expected.

**Comment by techee** [comment]
Yeah, "Auto" is the default to prevent Geany from freezing (tempararily) when opening huge projects. Good to hear it worked.

**Comment by elextr** [comment]
> If we wanted to support this feature, we'd have to modify quite a bit of the TM code because right now the presence of filename is used to distinguish global tags from non-global ones:

@techee What is done differently between a "global" tag and one with a filename other than being unable to goto since there is nowhere to goto to?  AFAIU everything is pretty much thrown in the mix for most tags/symbols actions if it comes from the same language (or C and C++ combined IIUC)?

I know that tagmangler format does not have a filename and Geany has to continue to handle that, but ctags format does and I thought that Geany read ctags files using ctags reader?  In which case wouldn't it read the filename?  And therefore wouldn't ctags tags files allow goto and such without parsing every file in a directory at startup using ProjectOrganiser?

**Comment by techee** [comment]
> @techee What is done differently between a "global" tag and one with a filename other than being unable to goto since there is nowhere to goto to? AFAIU everything is pretty much thrown in the mix for most tags/symbols actions if it comes from the same language (or C and C++ combined IIUC)?

Each non-global tag points to a TMSourceFile struct (corresponding to an open file in Geany):

https://github.com/geany/geany/blob/d54b2a899dca95639fa6f52d5e16034498c332d9/src/tagmanager/tm_tag.h#L101

which in turn contains the file name:

https://github.com/geany/geany/blob/d54b2a899dca95639fa6f52d5e16034498c332d9/src/tagmanager/tm_source_file.h#L32

TMSourceFile also contains all the tags appearing in the given file which are then merged into the global tag array. For global tags the `file` member of TMSourceFile is NULL. If we added TMSourceFile for global tags, we'd have to solve the following problems:

1. As I mentioned, NULL of the `file` member is used in TM to distinguish global tags from non-global tags.
2. The file name from a ctags tag file may not point to a valid path because it could have been generated on a different machine. It could even point to a file name which is already opened in Geany and for which TMSourceFile already exists but contains completely different tags.
3. Multiple ctags tag files possibly of different languages might point to the same path and TMSourceFile assumes it's from a single language.

I think the easiest way to solve this would be not doing anything with TMSourceFile and keeping it only for non-global tags and only adding a `filename` memeber of TMTag for global tags which could then be used for the goto.

**Comment by elextr** [comment]
> As I mentioned, NULL of the file member is used in TM to distinguish global tags from non-global tags.

So add a bool to the structure, mere matter of programming ;-P  

But my question was, apart from not having a filename, whats special about "global" tags that needs them to be distinguished?  Or is "global" just a confusing way of saying "no filename"?

Maybe needs an enum in TMSourceFile:

```
enum TMSourceTypeThingy { the_type_formerly_known_as_global, open_in_geany, tagfile_with_filename, future... };
```

> The file name from a ctags tag file may not point to a valid path because it could have been generated on a different machine. It could even point to a file name which is already opened in Geany and for which TMSourceFile already exists but contains completely different tags.

Ok, good point, but presumably the one open in Geany wins.  

And also the filename could point to a file of the same name but different contents, eg a different git checkout of the file.  I guess thats why etags format has the regexes to select the line, not just a hard wired line number.  Ooooh, _another_ tags format to play with [runs away fast].

> Multiple ctags tag files possibly of different languages might point to the same path and TMSourceFile assumes it's from a single language.

That would affect autocompletes, rarely is it good to call other languages functions directly, but it shouldn't impact goto, if a tag with the same language as the current file is loaded (using the extension of the filename stored in the TMSourceFile to decide), go there, else pop up the list and give the problem to the user.

Anyway, clearly its not something that is nearly there as I hoped it might be, oh well never mind.

**Comment by techee** [comment]
> But my question was, apart from not having a filename, whats special about "global" tags that needs them to be distinguished? Or is "global" just a confusing way of saying "no filename"?

In autocompletion, for instance, we put non-global tags first, see the sorting in

https://github.com/geany/geany/pull/3269

so we need to distinguish them somehow. And of course, we can use some boolean or, what I would prefer, having some `filename` member directly in TMTag for global tags.

> Ok, good point, but presumably the one open in Geany wins.

This is another point - Geany itself currently doesn't open files when doing goto definition/declaration (it's the ProjectOrganizer plugin which does for the files which are in the project directory). Geany assumes that all there exists are the open documents and only toggles among them. Should Geany itself start opening documents if it finds a matching global tag with a filename?

> That would affect autocompletes, rarely is it good to call other languages functions directly, but it shouldn't impact goto, if a tag with the same language as the current file is loaded (using the extension of the filename stored in the TMSourceFile to decide), go there, else pop up the list and give the problem to the user.

My point was that TMSourceFile assumes all its tags are from the same language and I'm not sure if it wouldn't cause some problems with tag management. I would personally avoid creating TMSourceFile for global tags and would prefer having some `filename` member of TMTag if we decided to implement it.

> Anyway, clearly its not something that is nearly there as I hoped it might be, oh well never mind.

It wouldn't be really that hard to implement, we just have to decide if we really want it and if Geany itself should start opening files when doing the goto lookup.

**Comment by elextr** [comment]
> Geany itself currently doesn't open files when doing goto definition/declaration (it's the ProjectOrganizer plugin which does for the files which are in the project directory)

> It wouldn't be really that hard to implement, we just have to decide if we really want it and if Geany itself should start opening files when doing the goto lookup.

I didn't examine the code carefully for when its used, but I noticed [goto-tag](https://github.com/geany/geany/blob/d54b2a899dca95639fa6f52d5e16034498c332d9/src/symbols.c#L1648) has [document_open](https://github.com/geany/geany/blob/d54b2a899dca95639fa6f52d5e16034498c332d9/src/symbols.c#L1696) and so does the popup_item_activate() when a list is displayed so I assumed that it did open the file, or at least there is an intention to open documents if the tag has a filename even if it doesn't work?  Of course that would be tags from ctags format files, not tagmangler files.

> I would personally avoid creating TMSourceFile for global tags and would prefer having some filename member of TMTag if we decided to implement it.

With all due respect to your adopted baby, I personally would prefer to throw the binary, Unicode dangerous, tagmanager format under a truck and use e/ctags format, even if it means using e/ctags to write the files.

**Comment by techee** [comment]
> I didn't examine the code carefully for when its used, but I noticed [goto-tag](https://github.com/geany/geany/blob/d54b2a899dca95639fa6f52d5e16034498c332d9/src/symbols.c#L1648) has [document_open](https://github.com/geany/geany/blob/d54b2a899dca95639fa6f52d5e16034498c332d9/src/symbols.c#L1696) and so does the popup_item_activate() when a list is displayed so I assumed that it did open the file, or at least there is an intention to open documents if the tag has a filename even if it doesn't work?

Alright, I was wrong then, Geany really opens the file. Anyway, we still need to store the path of global tags somewhere in the Geany's memory representation and it would be better to store it directly in TMTag than to TMSourceFile which is used for "dynamic" files like an editor that gets re-parsed every time the user writes something and is used for removing the original tags with the new tags. Global tags, instead, stay in the workspace unchanged since they get loaded.

> With all due respect to your adopted baby, I personally would prefer to throw the binary, Unicode dangerous, tagmanager format under a truck and use e/ctags format, even if it means using e/ctags to write the files.

I don't really have any problem with that and I think in principle we actually should do that, I'd just keep the old tag file format for compatibility (it's also useful for our unit tests because it reveals to which internal). I was also thinking we might add some basic escaping for characters `\200`-`\215` to avoid the unicode problems.

**Comment by elextr** [comment]
> I was also thinking we might add some basic escaping for characters \200-\215 to avoid the unicode problems.

Well, if its tagmanager format 2.0 all sorts of things are possible ...  ;-)

**Comment by techee** [comment]
> Well, if its tagmanager format 2.0 all sorts of things are possible ... ;-)

No, I'm talking about our current tagmanager 0.01alpha format. The characters from `\200`-`\215` may appear only once in the current format so we could use e.g. `\200\200` as the start of the escape sequence followed by the character we want to escape. Before reading, we'd unescape such sequences. Such a file would be backwards compatible - without any character to escape it would be identical to current format and if there's a character to escape, it's already broken with our format and the brokeness would be preserved.

**Comment by elextr** [comment]
> without any character to escape it would be identical to current format and if there's a character to escape, it's already broken with our format and the brokeness would be preserved.

Ok, that would fix the Unicode problem, and agree its important to preserve brokenness ;-D
